e2etest/newe2e_runazcopy_stdout.go

```diff
@@ -0,0 +1,198 @@
+	subscribers map[chan<- T]bool
+}
+
+func (m *ManySubscriberChannel[T]) Subscribe(target chan<- T) {
+	if m.subscribers == nil {
+		m.subscribers = make(map[chan<- T]bool)
+	}
+
+	m.subscribers[target] = true
+}
+
+func (m *ManySubscriberChannel[T]) SubscribeFunc(target func(T)) (handle chan<- T) {
+	targetChannel := make(chan T)
+
+	go func() {
+		for {
+			data, ok := <-targetChannel
+			if !ok {
+				break
+			}
+
+			target(data)
+		}
+	}()
+
+	m.Subscribe(targetChannel)
+	return targetChannel
+}
+
+// Unsubscribe will close the target channel, in addition to removing it from this channel's subscribers.
+func (m *ManySubscriberChannel[T]) Unsubscribe(target chan<- T) {
+	if m.subscribers == nil {
+		return
+	}
+
+	delete(m.subscribers, target)
+	close(target)
+}
+
+func (m *ManySubscriberChannel[T]) Message(data T) {
+	for k := range m.subscribers {
+		k <- data
+	}
+}
+
+// AzCopyParsedStdout is still a semi-raw stdout struct.
+type AzCopyParsedStdout struct {
+	Messages     []common.JsonOutputTemplate
+	OnParsedLine ManySubscriberChannel[common.JsonOutputTemplate]
+}
+
+func (a *AzCopyParsedStdout) RawStdout() []string {
+	out := make([]string, len(a.Messages))
+
+	for k, v := range a.Messages {
+		buf, _ := json.Marshal(v)
+
+		out[k] = string(buf)
+	}
+
+	return out
+}
+
+func (a *AzCopyParsedStdout) Write(p []byte) (n int, err error) {
+	str := string(p)
+	lines := strings.Split(strings.TrimSuffix(str, "\n"), "\n")
+	n = len(p)
+
+	for _, v := range lines {
+		var out common.JsonOutputTemplate
+		err = json.Unmarshal([]byte(v), &out)
+		if err != nil {
+			return
+		}
+
+		a.OnParsedLine.Message(out)
+		a.Messages = append(a.Messages, out)
+	}
+
+	return
+}
+
+func (a *AzCopyParsedStdout) String() string {
+	return strings.Join(a.RawStdout(), "\n")
+}
+
+type AzCopyParsedListStdout struct {
+	AzCopyParsedStdout
+	listenChan chan<- common.JsonOutputTemplate
+
+	Items   map[AzCopyOutputKey]cmd.AzCopyListObject
+	Summary cmd.AzCopyListSummary
+}
+
+func (a *AzCopyParsedListStdout) InsertObject(obj cmd.AzCopyListObject) {
+	versionName := common.Iff(obj.VersionId == "", "latest", obj.VersionId)
+	var parsedVersionID time.Time
+	if versionName == "latest" {
+		parsedVersionID = DerefOrZero(obj.LastModifiedTime)
+	} else {
+		parsedVersionID, _ = time.Parse(time.RFC3339, obj.VersionId)
+	}
+
+	a.Items[AzCopyOutputKey{
+		Path:      obj.Path,
+		VersionId: parsedVersionID.Format(time.RFC3339),
+	}] = cmd.AzCopyListObject{}
+}
+
+func (a *AzCopyParsedListStdout) Write(p []byte) (n int, err error) {
+	if a.listenChan == nil {
+		a.listenChan = a.OnParsedLine.SubscribeFunc(func(line common.JsonOutputTemplate) {
+			switch line.MessageType {
+			case "ListObject":
+				var object cmd.AzCopyListObject
+				err = json.Unmarshal([]byte(line.MessageContent), &object)
+				if err != nil {
+					return
+				}
+
+				a.InsertObject(object)
+
+			case "ListSummary":
+				_ = json.Unmarshal([]byte(line.MessageContent), &a.Summary)
```

Comment:
should we catch an error here?

Reply:
Probably. Actually, there's room to suggest that "probably not", because if writing fails, the exec.Cmd stops running.